Cython/Compiler/PyrexTypes.py

```diff
@@ -4599,12 +4599,22 @@ def best_match(arg_types, functions, pos=None, env=None, args=None):
             errors.append((func, error_mesg))
             continue
         if func_type.templates:
+            # For any argument/parameter pair A/P, if P is a forwarding reference,
+            # use lvalue-reference-to-A for deduction in place of A when the
+            # function call argument is an lvalue. See:
+            # https://en.cppreference.com/w/cpp/language/template_argument_deduction#Deduction_from_a_function_call
+            arg_types_for_deduction = arg_types.copy()
+            if func.type.is_cfunction and args:
+                for i, formal_arg in enumerate(func.type.args):
+                    if formal_arg.is_forwarding_reference():
+                        if args[i].is_lvalue():
```

Comment:
Worth writing tests for (static/normal) methods since that changes `args`. The original version strictly uses `self.args`.

Reply:
To clarify, do you mean additional tests in `cpp_forwarding_ref.pyx` for templated _methods_ in addition to free functions? Apologies, it's been a while.. 🙂

Reply:
I went ahead and tried that. Normal methods seem to work as expected, but overloaded static methods don't. Looking into why...

Reply:
Hmm - it could be unrelated to this PR, since the following fails on `master` for me:

```cython
# mode: run
# tag: cpp, cpp11

cdef extern from *:
    """
    struct Foo
    {
      static const char* bar(int x, int y) {
        return "second";
      }

      static const char* bar(int x) {
        return "first";
      }

    };
    """
    cppclass Foo:
        @staticmethod
        const char* bar(int x)


def test_static_method_overload():
    """
    >>> test_static_method_overload()
    """
    assert Foo.bar(1) == b"first"
    assert Foo.bar(1, 2) == b"second"
```

Error:

```
=== Expected: ===


=== Got: ===
28:18: Call with wrong number of arguments (expected 1, got 2)

======================================================================
ERROR: runTest (__main__.CythonRunTestCase)
compiling (cpp/cy2) and running cpp_static_method_overload
----------------------------------------------------------------------
Traceback (most recent call last):
  File "runtests.py", line 1312, in run
    ext_so_path = self.runCompileTest()
  File "runtests.py", line 977, in runCompileTest
    return self.compile(
  File "runtests.py", line 1233, in compile
    self._match_output(expected_errors, errors, tostderr)
  File "runtests.py", line 1285, in _match_output
    self.assertEqual(None, unexpected)
AssertionError: None != '28:18: Call with wrong number of arguments (expected 1, got 2)'

----------------------------------------------------------------------
```

Reply:
That looks known https://github.com/cython/cython/issues/1851 so maybe isn't with worrying about too much

Reply:
Thanks! I got curious and opened https://github.com/cython/cython/pull/4129 to address that :-)